### ISSUE TITLE
`clippy::not_unsafe_ptr_arg_deref` false positive

### ISSUE THREAD

**Issue by VVishion** [rust-lang/rust-clippy#17193]
### Summary

Im suspecting the following to be a false positive. In my head it should never be able to cause UB, because these functions never forms a reference and just computes a new raw ptr with offset address.

```rust
    // does not form a reference does therefore not cause UB
    pub const fn as_mut_ptr_raw(this: *mut Self) -> *mut T {
        let storage = unsafe { &raw mut (*this).0.storage };
        storage as *mut T
    }

    // does not form a reference does therefore not cause UB
    pub const fn as_ptr_raw(this: *const Self) -> *const T {
        let storage = unsafe { &raw const (*this).0.storage };
        storage as *const T
    }
```

### Lint Name

not_unsafe_ptr_arg_deref

### Reproducer

I tried this code:

```rust
use std::mem::MaybeUninit;

pub struct Buf<T>(pub(crate) Container<[MaybeUninit<T>]>);

pub(crate) struct Container<Storage: ?Sized> {
    pub len: usize,
    pub storage: Storage,
}

impl<T> Buf<T> {
    // does not form a reference does therefore not cause UB
    pub const fn as_mut_ptr_raw(this: *mut Self) -> *mut T {
        let storage = unsafe { &raw mut (*this).0.storage };
        storage as *mut T
    }

    // does not form a reference does therefore not cause UB
    pub const fn as_ptr_raw(this: *const Self) -> *const T {
        let storage = unsafe { &raw const (*this).0.storage };
        storage as *const T
    }
}
```

I saw this happen:

```
error: this public function might dereference a raw pointer but is not marked `unsafe`
  --> src/lib.rs:13:43
   |
13 |         let storage = unsafe { &raw mut (*this).0.storage };
   |                                           ^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.96.0/index.html#not_unsafe_ptr_arg_deref
   = note: `#[deny(clippy::not_unsafe_ptr_arg_deref)]` on by default

error: this public function might dereference a raw pointer but is not marked `unsafe`
  --> src/lib.rs:19:45
   |
19 |         let storage = unsafe { &raw const (*this).0.storage };
   |                                             ^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.96.0/index.html#not_unsafe_ptr_arg_deref

```

I expected to see this happen:
no complaints

### Version

```text
rustc 1.98.0-nightly (f20a92ec0 2026-06-07)
binary: rustc
commit-hash: f20a92ec01483dc5c58e90e246f266bdad822d86
commit-date: 2026-06-07
host: x86_64-unknown-linux-gnu
release: 1.98.0-nightly
LLVM version: 22.1.6
```

**Comment by y21** [comment]
Field offset calculation itself still has safety requirements, that's why field projection on `&raw` still requires an unsafe block. The safety section for [`core::ptr::addr_of!`](https://doc.rust-lang.org/stable/std/ptr/macro.addr_of.html#safety) explains this bit in more detail. So these functions are still unsound when e.g. passed a dangling pointer.

**Comment by VVishion** [comment]
I see. Thank you
